File: petroka_custom/petroka_custom/report/task_summary_version_i/task_summary_version_i.js
// Copyright (c) 2026, Administrator and contributors
// For license information, please see license.txt

frappe.query_reports["Task Summary version-I"] = {
	filters: [
		{
			fieldname: "start_date",
			label: __("Start Date"),
			fieldtype: "Date",
		},
		{
			fieldname: "end_date",
			label: __("End Date"),
			fieldtype: "Date",
		},
		{
			fieldname: "employee",
			label: __("Employee"),
			fieldtype: "Link",
			options: "Employee",

			get_query() {
				const department =
					frappe.query_report.get_filter_value(
						"department"
					);

				const filters = {
					status: "Active",
				};

				if (department) {
					filters.department = department;
				}

				return {
					filters: filters,
				};
			},
		},
		{
			fieldname: "department",
			label: __("Department"),
			fieldtype: "Link",
			options: "Department",
		},
		{
			fieldname: "project",
			label: __("Project"),
			fieldtype: "Link",
			options: "Project",

			get_query() {
				const project_statuses =
					frappe.query_report.get_filter_value(
						"project_status"
					);

				const statuses = parse_multiselect_value(
					project_statuses
				);

				if (!statuses.length) {
					return {};
				}

				return {
					filters: {
						status: ["in", statuses],
					},
				};
			},
		},
		{
			fieldname: "task",
			label: __("Task"),
			fieldtype: "Link",
			options: "Task",

			get_query() {
				const project =
					frappe.query_report.get_filter_value(
						"project"
					);

				const task_statuses =
					frappe.query_report.get_filter_value(
						"task_status"
					);

				const statuses = parse_multiselect_value(
					task_statuses
				);

				const filters = {
					is_template: 0,
				};

				if (project) {
					filters.project = project;
				}

				if (statuses.length) {
					filters.status = ["in", statuses];
				}

				return {
					filters: filters,
				};
			},
		},
		{
			fieldname: "project_status",
			label: __("Project Status"),
			fieldtype: "MultiSelectList",
			default: [
				"Open",
				"In Progress",
				"On Hold",
			],

			get_data(txt) {
				return get_status_options(
					[
						"Open",
						"In Progress",
						"On Hold",
						"Completed",
						"Cancelled",
					],
					txt
				);
			},
		},
		{
			fieldname: "task_status",
			label: __("Task Status"),
			fieldtype: "MultiSelectList",
			default: [
				"Open",
				"Working",
				"Pending Review",
				"Overdue",
				"To be Done",
			],

			get_data(txt) {
				return get_status_options(
					[
						"Open",
						"Working",
						"Pending Review",
						"Overdue",
						"To be Done",
						"Completed",
						"Cancelled",
						"Template",
					],
					txt
				);
			},
		},
	],

	onload(report) {
		setup_filter_dependencies(report);
	},
};


function get_status_options(statuses, txt) {
	const search_text = (txt || "")
		.toLowerCase()
		.trim();

	return statuses
		.filter((status) => {
			return status
				.toLowerCase()
				.includes(search_text);
		})
		.map((status) => {
			return {
				value: status,
				description: status,
			};
		});
}


function parse_multiselect_value(value) {
	if (!value) {
		return [];
	}

	if (Array.isArray(value)) {
		return value;
	}

	if (typeof value === "string") {
		try {
			const parsed_value = JSON.parse(value);

			if (Array.isArray(parsed_value)) {
				return parsed_value;
			}
		}
		catch (error) {
			return value
				.split(",")
				.map((item) => item.trim())
				.filter(Boolean);
		}
	}

	return [];
}


function setup_filter_dependencies(report) {
	const department_filter =
		report.get_filter("department");

	const project_filter =
		report.get_filter("project");

	if (department_filter) {
		department_filter.$input.on(
			"change",
			() => {
				report.set_filter_value(
					"employee",
					""
				);
			}
		);
	}

	if (project_filter) {
		project_filter.$input.on(
			"change",
			() => {
				report.set_filter_value(
					"task",
					""
				);
			}
		);
	}
}

File: petroka_custom/petroka_custom/report/task_summary_version_i/task_summary_version_i.py
# Copyright (c) 2026, Administrator and contributors
# For license information, please see license.txt

import json

import frappe
from frappe import _


def execute(filters=None):
	filters = frappe._dict(filters or {})

	columns = get_columns()
	data = get_data(filters)

	return columns, data


def get_columns():
	return [
		{
			"fieldname": "project",
			"label": _("Project"),
			"fieldtype": "Link",
			"options": "Project",
			"width": 200,
		},
		{
			"fieldname": "task",
			"label": _("Task"),
			"fieldtype": "Link",
			"options": "Task",
			"width": 160,
		},
		{
			"fieldname": "subject",
			"label": _("Subject"),
			"fieldtype": "Data",
			"width": 280,
		},
		{
			"fieldname": "employee",
			"label": _("Employee"),
			"fieldtype": "Link",
			"options": "Employee",
			"width": 180,
		},
		{
			"fieldname": "expected_hours",
			"label": _("Expected Hours"),
			"fieldtype": "Float",
			"precision": 3,
			"width": 130,
		},
		{
			"fieldname": "actual_hours",
			"label": _("Actual Hours"),
			"fieldtype": "Float",
			"precision": 3,
			"width": 120,
		},
		{
			"fieldname": "employee_working_hours",
			"label": _("Employee Working Hours"),
			"fieldtype": "Float",
			"precision": 3,
			"width": 170,
		},
		{
			"fieldname": "remaining_hours",
			"label": _("Remaining Hours"),
			"fieldtype": "Float",
			"precision": 3,
			"width": 140,
		},
		{
			"fieldname": "total_costing_amount",
			"label": _("Total Costing Amount"),
			"fieldtype": "Float",
			"precision": 2,
			"width": 170,
		},
		{
			"fieldname": "total_billing_amount",
			"label": _("Total Billable Amount"),
			"fieldtype": "Float",
			"precision": 2,
			"width": 170,
		},
		{
			"fieldname": "department",
			"label": _("Department"),
			"fieldtype": "Link",
			"options": "Department",
			"width": 180,
			"hidden": 1,
		},
		{
			"fieldname": "company",
			"label": _("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"width": 180,
			"hidden": 1,
		},
		{
			"fieldname": "currency",
			"label": _("Currency"),
			"fieldtype": "Data",
			"hidden": 1,
		},
	]


def get_data(filters):
	conditions = [
		"task.project IS NOT NULL",
		"COALESCE(task.is_template, 0) = 0",
	]

	values = {}

	add_date_filters(filters, conditions, values)
	add_link_filters(filters, conditions, values)
	add_status_filters(filters, conditions, values)

	condition_string = " AND ".join(conditions)

	return frappe.db.sql(
		f"""
		SELECT
			task.project AS project,
			task.name AS task,
			task.subject AS subject,

			employee.name AS employee,
			employee.department AS department,

			COALESCE(
				task.expected_time,
				0
			) AS expected_hours,

			COALESCE(
				task.actual_time,
				0
			) AS actual_hours,

			COALESCE(
				employee_hours.working_hours,
				0
			) AS employee_working_hours,

			(
				COALESCE(task.expected_time, 0)
				-
				COALESCE(task.actual_time, 0)
			) AS remaining_hours,

			COALESCE(
				task.total_costing_amount,
				0
			) AS total_costing_amount,

			COALESCE(
				task.total_billing_amount,
				0
			) AS total_billing_amount,

			project.company AS company,
			company.default_currency AS currency

		FROM `tabTask` AS task

		LEFT JOIN `tabProject` AS project
			ON project.name = task.project

		LEFT JOIN `tabCompany` AS company
			ON company.name = project.company

		LEFT JOIN (
			SELECT DISTINCT
				reference_name,
				allocated_to

			FROM `tabToDo`

			WHERE
				reference_type = 'Task'
				AND status != 'Cancelled'
		) AS assignment
			ON assignment.reference_name = task.name

		LEFT JOIN `tabEmployee` AS employee
			ON employee.user_id = assignment.allocated_to

		LEFT JOIN (
			SELECT
				timesheet.employee AS employee,
				detail.task AS task,
				SUM(
					COALESCE(detail.hours, 0)
				) AS working_hours

			FROM `tabTimesheet Detail` AS detail

			INNER JOIN `tabTimesheet` AS timesheet
				ON timesheet.name = detail.parent

			WHERE
				timesheet.docstatus = 1
				AND detail.task IS NOT NULL
				AND detail.task != ''

			GROUP BY
				timesheet.employee,
				detail.task
		) AS employee_hours
			ON employee_hours.task = task.name
			AND employee_hours.employee = employee.name

		WHERE {condition_string}

		ORDER BY
			task.project ASC,
			task.exp_start_date ASC,
			task.exp_end_date ASC,
			task.name ASC,
			employee.employee_name ASC
		""",
		values=values,
		as_dict=True,
	)


def add_date_filters(filters, conditions, values):
	if filters.get("start_date"):
		conditions.append(
			"task.exp_start_date >= %(start_date)s"
		)
		values["start_date"] = filters.start_date

	if filters.get("end_date"):
		conditions.append(
			"task.exp_end_date <= %(end_date)s"
		)
		values["end_date"] = filters.end_date


def add_link_filters(filters, conditions, values):
	if filters.get("employee"):
		conditions.append(
			"employee.name = %(employee)s"
		)
		values["employee"] = filters.employee

	if filters.get("department"):
		conditions.append(
			"employee.department = %(department)s"
		)
		values["department"] = filters.department

	if filters.get("project"):
		conditions.append(
			"task.project = %(project)s"
		)
		values["project"] = filters.project

	if filters.get("task"):
		conditions.append(
			"task.name = %(task)s"
		)
		values["task"] = filters.task


def add_status_filters(filters, conditions, values):
	project_statuses = get_multiselect_values(
		filters.get("project_status")
	)

	task_statuses = get_multiselect_values(
		filters.get("task_status")
	)

	if project_statuses:
		conditions.append(
			"project.status IN %(project_statuses)s"
		)
		values["project_statuses"] = tuple(project_statuses)

	if task_statuses:
		conditions.append(
			"task.status IN %(task_statuses)s"
		)
		values["task_statuses"] = tuple(task_statuses)


def get_multiselect_values(value):
	if not value:
		return []

	if isinstance(value, (list, tuple)):
		return [
			item
			for item in value
			if item
		]

	if isinstance(value, str):
		try:
			parsed_value = json.loads(value)

			if isinstance(parsed_value, list):
				return [
					item
					for item in parsed_value
					if item
				]

		except (json.JSONDecodeError, TypeError):
			pass

		return [
			item.strip()
			for item in value.split(",")
			if item.strip()
		]

	return []